README.md
# Tetra

[![Build Status](https://img.shields.io/github/workflow/status/17cupsofcoffee/tetra/CI%20Build/main)](https://github.com/17cupsofcoffee/tetra/actions?query=branch%3Amain)
[![Crates.io](https://img.shields.io/crates/v/tetra.svg)](https://crates.io/crates/tetra)
[![Documentation](https://docs.rs/tetra/badge.svg)](https://docs.rs/tetra)
[![License](https://img.shields.io/crates/l/tetra.svg)](LICENSE)

Tetra is a simple 2D game framework written in Rust. It uses SDL2 for event handling and OpenGL 3.2+ for rendering.

* [Website](https://tetra.seventeencups.net)
* [Tutorial](https://tetra.seventeencups.net/tutorial)
* [API Docs](https://docs.rs/tetra)
* [FAQ](https://tetra.seventeencups.net/FAQ)

## Features

* XNA/MonoGame-inspired API
* Efficient 2D rendering, with draw call batching by default
* Easy input handling, via polling or events, with support for gamepads
* Deterministic game loop by default, à la [Fix Your Timestep](https://gafferongames.com/post/fix_your_timestep/)
* Common building blocks built-in, such as:
    * Font rendering
    * Cameras
    * Screen scaling

## Installation

To add Tetra to your project, add the following line to your `Cargo.toml` file:

```toml
tetra = "0.6"
```

You will also need to install the SDL2 native libraries - full details are provided in the [documentation](https://tetra.seventeencups.net/installation).

## Examples

To get a simple window displayed on screen, the following code can be used:

```rust ,noplaypen
use firecore_tetra::graphics::{self, Color};
use firecore_tetra::{Context, ContextBuilder, State};

struct GameState;

impl State for GameState {
    fn draw(&mut self, ctx: &mut DefaultContext) -> firecore_tetra::Result {
        // Cornflower blue, as is tradition
        graphics::clear(ctx, Color::rgb(0.392, 0.584, 0.929));
        Ok(())
    }
}

fn main() -> firecore_tetra::Result {
    ContextBuilder::new("Hello, world!", 1280, 720)
        .build()?
        .run(|_| Ok(GameState))
}
```

You can see this example in action by running `cargo run --example hello_world`.

The full list of examples is available [here](https://tetra.seventeencups.net/examples).

## Support/Feedback

Tetra is fairly early in development, so you might run into bugs/flaky docs/general weirdness. Please feel free to open an issue/PR if you find something! You can also contact me via [Twitter](https://twitter.com/17cupsofcoffee) or the [Rust Game Development Discord](https://discord.gg/yNtPTb2).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file, following the format defined at keepachangelog.com. Where a change was contributed via a third-party pull request, the author will be credited.

From 0.4.0 onwards, all breaking changes will be explicitly labelled, to make it easier to assess the impact of upgrading.

This project adheres to Semantic Versioning.

## [0.6.5] - 2021-05-26

### Added 

* Uniform arrays are now supported in shaders, and can be set by passing an array or a slice to `set_uniform`.
* Basic support for instanced mesh rendering has been added, via `Mesh::draw_instanced`.

### Fixed

* Switching away from a non-multisampled canvas no longer resolves the canvas.
    * This was a regression due to a refactor in 0.6.3 - it should not have caused any change to the rendered image, but may have negatively impacted performance.

## [0.6.4] - 2021-05-14

### Added

* Stencil buffers can now be used, both with the main framebuffer and with `Canvas` rendering. ([@tesselode](https://github.com/tesselode) in [#261](https://github.com/17cupsofcoffee/tetra/pull/261))
    * Stencil buffers will only be created if they are requested via `ContextBuilder::stencil_buffer` or `CanvasBuilder::stencil_buffer`.
* `window::get_physical_size` has been added, making it possible to find out the size of the window in unscaled, physical pixels.
* `window::get_dpi_scale` has been added, allowing the window/monitor's pixel density to be queried. ([@sumibi-yakitori](https://github.com/sumibi-yakitori) in [#249](https://github.com/17cupsofcoffee/tetra/pull/249))
* `Texture` and `Canvas` now have a `get_data` method, allowing their texture data to be copied from the GPU into an `ImageData`.
* `ImageData` now has `as_bytes` and `into_bytes` method, which allows you to access the underlying buffer.

### Changed

* Canvases are now initialized to transparent black, rather than containing unspecified data on creation.
* Multisampled canvases will now automatically lower their level of multisampling (or disable it altogether) if the hardware does not support the requested number of samples.
    * This should allow games with multisampling to gracefully degrade on older/weaker GPUs.
* `ImageData::from_rgba` has been replaced with `ImageData::from_rgba8`, and `BmFontBuilder::with_page_rgba` has been replaced with `BmFontBuilder::with_page_rgba8`.
    * This is for consistency with existing methods (e.g. on `Color`).
    * The old names have been deprecated, and will be removed in 0.7.
* Updated `glow` to 0.9.

### Fixed

* `Texture::filter_mode` and `Canvas::filter_mode` now return the correct initial value. Previously, they would always return `FilterMode::Linear` until `set_filter_mode` was called. 
* DPI scaling should now continue to work after calling `graphics::reset_canvas`.
* `graphics::set_scissor` is now properly aware of DPI scaling.
* When a texture, canvas or buffer fails to be created by the GPU (e.g. due to being out of memory or going over the max size), an error will now be returned, instead of unspecified behaviour occurring.

### Deprecated

* `ImageData::from_rgba` and `BmFontBuilder::with_page_rgba` has been deprecated in favour of more consistently named functions (see above).
* `Canvas::multisampled` has been deprecated in favour of `Canvas::builder`.

## [0.6.3] - 2021-04-09

### Added 

* BMFont-formatted fonts are now supported, via `Font::bmfont` and `BmFontBuilder`.
* Various `std` operator traits have been implemented for the combination of `Color` and `f32`, allowing for all four components to be operated on at once.
    * This can be especially useful when working with premultiplied alpha - to set the opacity of a opaque color, you can now just multiply it by an `f32`.
* `Color` now has a `to_premultiplied` method, which can be used to convert a color into its premultiplied version.
* `Color` can now be converted to and from `[f32; 4]` and `[u8; 4]` via `From` and `Into`.
* The `ImageData` struct can be used to load and manipulate image data on the CPU, without the need for a `Context`.
    * This can be useful for asset loading/pre-processing.
    * To avoid future breaking changes if/when Tetra gains support for multiple pixel formats, the API doesn't currently give access to the underlying raw pixel data. It instead operates purely on `Color`s, automatically converting to and from as needed.
* `Color` and `Vertex` now implement `Default`, with the default values being all zeroes (for consistency with the vector types, and to make them more useful for zeroing out buffers).

### Changed

* Improved docs for the `UniformValue` trait.

## [0.6.2] - 2021-03-15

### Fixed

* `VertexBuffer::set_data` was mistakenly measuring its offset in individual floats, rather than vertices. This was inconsistent with `IndexBuffer`, and could potentially lead to corrupted data.
    * I do not believe this was a memory safety issue, as all writes were still valid and aligned - they were just in the wrong place!

## [0.6.1] - 2021-03-15

### Added

* Blend modes can now be set when rendering (including variants for pre-multiplied alpha). ([@tesselode](https://github.com/tesselode) in [#244](https://github.com/17cupsofcoffee/tetra/pull/244)) 
* Scissor rectangles can now be set when rendering, via `graphics::set_scissor` and `graphics::reset_scissor`. 
* `Rectangle::combine` has been added, which can be used to get a rectangle that contains two other rectangles.
* Fonts can now have their filter mode set, instead of always using nearest neighbor filtering. ([@tesselode](https://github.com/tesselode) in [#246](https://github.com/17cupsofcoffee/tetra/pull/246)) 
    * Fonts now also respect the default filter mode when they are created.
* Text now supports word-wrapping, via the `Text::wrapped` constructor or the `Text::set_max_width` method.

src/lib.rs
//! Tetra is a simple 2D game framework written in Rust. It uses SDL2 for event handling and OpenGL 3.2+ for rendering.
//!
//! * [Website](https://tetra.seventeencups.net)
//! * [Tutorial](https://tetra.seventeencups.net/tutorial)
//! * [API Docs](https://docs.rs/tetra)
//! * [FAQ](https://tetra.seventeencups.net/FAQ)
//!
//! ## Features
//!
//! * XNA/MonoGame-inspired API
//! * Efficient 2D rendering, with draw call batching by default
//! * Easy input handling, via polling or events, with support for gamepads
//! * Deterministic game loop by default, à la [Fix Your Timestep](https://gafferongames.com/post/fix_your_timestep/)
//! * Common building blocks built-in, such as:
//!     * Font rendering
//!     * Cameras
//!     * Screen scaling
//!
//! ## Installation
//!
//! To add Tetra to your project, add the following line to your `Cargo.toml` file:
//!
//! ```toml
//! tetra = "0.6"
//! ```
//!
//! You will also need to install the SDL2 native libraries - full details are provided in the [documentation](https://tetra.seventeencups.net/installation).
//!
//! ## Examples
//!
//! To get a simple window displayed on screen, the following code can be used:
//!
//! ```no_run
//! use firecore_tetra::graphics::{self, Color};
//! use firecore_tetra::{Context, ContextBuilder, DefaultContext, State};
//!
//! struct GameState;
//!
//! impl State for GameState {
//!     fn draw(&mut self, ctx: &mut DefaultContext) -> firecore_tetra::Result {
//!         // Cornflower blue, as is tradition
//!         graphics::clear(ctx, Color::rgb(0.392, 0.584, 0.929));
//!         Ok(())
//!     }
//! }
//!
//! fn main() -> firecore_tetra::Result {
//!     ContextBuilder::new("Hello, world!", 1280, 720)
//!         .build()?
//!         .run(|_| Ok(GameState))
//! }
//! ```
//!
//! You can see this example in action by running `cargo run --example hello_world`.
//!
//! The full list of examples is available [here](https://tetra.seventeencups.net/examples).
//!
//! ## Support/Feedback
//!
//! Tetra is fairly early in development, so you might run into bugs/flaky docs/general weirdness. Please feel free to open an issue/PR if you find something! You can also contact me via [Twitter](https://twitter.com/17cupsofcoffee) or the [Rust Game Development Discord](https://discord.gg/yNtPTb2).

#![warn(missing_docs)]

#[cfg(feature = "audio")]
pub mod audio;
mod context;
pub mod error;
mod fs;
pub mod graphics;
pub mod input;
mod lifecycle;
pub mod math;
mod platform;
pub mod time;
pub mod window;

pub use crate::context::{run, Context, ContextBuilder, DefaultContext};
pub use crate::error::{Result, TetraError};
pub use crate::lifecycle::{Event, State};
